File: rtp_llm/cpp/model_rpc/test/RecentCacheKeyWindowTest.cc
#include "rtp_llm/cpp/model_rpc/RecentCacheKeyWindow.h"

#include <atomic>
#include <cstdlib>

#include "gtest/gtest.h"

namespace rtp_llm {
namespace {

TEST(RecentCacheKeyWindowTest, HitIsComputedBeforeInsert) {
    std::atomic<int64_t> now{1000};
    RecentCacheKeyWindow window(1000, [&now]() { return now.load(); });

    auto first = window.record({1, 2, 3});
    EXPECT_EQ(first.request_occurrences, 3);
    EXPECT_EQ(first.request_hit_occurrences, 0);
    EXPECT_DOUBLE_EQ(first.request_hit_ratio, 0.0);
    EXPECT_EQ(first.retained_occurrences, 3);
    EXPECT_EQ(first.retained_unique_cache_keys, 3);

    auto second = window.record({2, 3, 4});
    EXPECT_EQ(second.request_occurrences, 3);
    EXPECT_EQ(second.request_hit_occurrences, 2);
    EXPECT_DOUBLE_EQ(second.request_hit_ratio, 2.0 / 3.0);
    EXPECT_EQ(second.retained_occurrences, 6);
    EXPECT_EQ(second.retained_unique_cache_keys, 4);
}

TEST(RecentCacheKeyWindowTest, DuplicateKeysCountByOccurrence) {
    std::atomic<int64_t> now{1000};
    RecentCacheKeyWindow window(1000, [&now]() { return now.load(); });

    auto first = window.record({7, 7, 7});
    EXPECT_EQ(first.request_occurrences, 3);
    EXPECT_EQ(first.request_hit_occurrences, 0);
    EXPECT_EQ(first.retained_occurrences, 3);
    EXPECT_EQ(first.retained_unique_cache_keys, 1);

    auto second = window.record({7, 7});
    EXPECT_EQ(second.request_occurrences, 2);
    EXPECT_EQ(second.request_hit_occurrences, 2);
    EXPECT_EQ(second.retained_occurrences, 5);
    EXPECT_EQ(second.retained_unique_cache_keys, 1);
}

TEST(RecentCacheKeyWindowTest, ExpiredEntriesDecrementCountsAndRemoveZeroKeys) {
    std::atomic<int64_t> now{1000};
    RecentCacheKeyWindow window(1000, [&now]() { return now.load(); });

    window.record({1, 2});
    now = 1500;
    window.record({2, 3});
    auto snapshot = window.snapshot();
    EXPECT_EQ(snapshot.retained_occurrences, 4);
    EXPECT_EQ(snapshot.retained_unique_cache_keys, 3);

    now = 2001;
    snapshot = window.snapshot();
    EXPECT_EQ(snapshot.retained_occurrences, 2);
    EXPECT_EQ(snapshot.retained_unique_cache_keys, 2);

    auto next = window.record({1, 2, 3});
    EXPECT_EQ(next.request_hit_occurrences, 2);
    EXPECT_EQ(next.retained_occurrences, 5);
    EXPECT_EQ(next.retained_unique_cache_keys, 3);
}

TEST(RecentCacheKeyWindowTest, EmptyRequestReportsZeroButKeepsWindow) {
    RecentCacheKeyWindow window(1000, []() { return 1000; });
    window.record({1, 2});
    auto empty = window.record({});
    EXPECT_EQ(empty.request_occurrences, 0);
    EXPECT_EQ(empty.request_hit_occurrences, 0);
    EXPECT_EQ(empty.retained_occurrences, 2);
    EXPECT_EQ(empty.retained_unique_cache_keys, 2);
}

}  // namespace
}  // namespace rtp_llm

File: rtp_llm/server/test/recent_cache_key_window_test.py
import os
import unittest
from unittest.mock import patch

from rtp_llm.server.recent_cache_key_window import (
    CACHE_HIT_TIME_WINDOW_MS_ENV,
    DEFAULT_TIME_WINDOW_MS,
    RecentCacheKeyWindow,
)


class TestRecentCacheKeyWindow(unittest.TestCase):
    def test_counts_request_hits_against_prior_window(self):
        now_ms = 0
        window = RecentCacheKeyWindow(1000, lambda: now_ms)

        first = window.record([1, 2, 3])
        self.assertEqual(first.request_occurrences, 3)
        self.assertEqual(first.request_hit_occurrences, 0)
        self.assertEqual(first.request_hit_ratio, 0.0)

        second = window.record([2, 3, 4])
        self.assertEqual(second.request_occurrences, 3)
        self.assertEqual(second.request_hit_occurrences, 2)
        self.assertAlmostEqual(second.request_hit_ratio, 2 / 3)
        self.assertEqual(second.retained_occurrences, 6)
        self.assertEqual(second.retained_unique_cache_keys, 4)

    def test_expires_old_entries_before_matching(self):
        now_ms = 0

        def get_now_ms():
            return now_ms

        window = RecentCacheKeyWindow(1000, get_now_ms)
        window.record([1, 2, 3])
        now_ms = 1001

        snapshot = window.record([1, 4])
        self.assertEqual(snapshot.request_occurrences, 2)
        self.assertEqual(snapshot.request_hit_occurrences, 0)
        self.assertEqual(snapshot.retained_occurrences, 2)
        self.assertEqual(snapshot.retained_unique_cache_keys, 2)

    def test_repeated_keys_do_not_self_hit_within_one_request(self):
        now_ms = 0
        window = RecentCacheKeyWindow(1000, lambda: now_ms)

        first = window.record([7, 7, 7])
        self.assertEqual(first.request_occurrences, 3)
        self.assertEqual(first.request_hit_occurrences, 0)

        second = window.record([7, 7])
        self.assertEqual(second.request_occurrences, 2)
        self.assertEqual(second.request_hit_occurrences, 2)
        self.assertEqual(second.request_hit_ratio, 1.0)

    def test_uses_environment_time_window(self):
        with patch.dict(os.environ, {CACHE_HIT_TIME_WINDOW_MS_ENV: "1234"}):
            self.assertEqual(RecentCacheKeyWindow().time_window_ms, 1234)

    def test_invalid_environment_time_window_falls_back(self):
        with patch.dict(os.environ, {CACHE_HIT_TIME_WINDOW_MS_ENV: "bad"}):
            self.assertEqual(RecentCacheKeyWindow().time_window_ms, DEFAULT_TIME_WINDOW_MS)


if __name__ == "__main__":
    unittest.main()
